### PR TITLE
[quant] remove reduce_range in customized qrange

### PR DESCRIPTION
**Pull request by terrychenism**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75357

Summary:
reduce_range in customized qrange is not needed, since user should set the customized qrange correctly
1. previously user can set customized qmin/max and reduce_range, but 'reduce_range' is not useful when user can customized the qconfig. Now we remove the reduce_range flag to let user use qmin/max only. 
2. after PR, user should take care of qmin/qmax for correct range, e.g. old qconfig ( qmin= 0, qmax = 255, reduce_range = true) , new qconfig( qmin = 0, qmax = 127) 
3.  BC fix:  quant_min, quant_max = quant_min // 2, quant_max // 2 when old qconfig contain reduce_range = True. 

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35452674](https://our.internmc.facebook.com/intern/diff/D35452674)

### CODE REVIEW

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35452674).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35452674).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35452674).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35452674).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35452674).

**Review by vkuzo — CHANGES_REQUESTED** [review]
Are we sure this is correct?  This PR seems to leave the reduce_range argument in the API, but changes it to be unused, which would make any customer who is actually trying to use this argument get incorrect results.  

What is the goal of this PR?

**Comment by terrychenism** [comment]
> correct



> Are we sure this is correct? This PR seems to leave the reduce_range argument in the API, but changes it to be unused, which would make any customer who is actually trying to use this argument get incorrect results.
> 
> What is the goal of this PR?

we delete reduce_range argument which in the custimized qconfig. if user use customized qmin/qmax, it means they can specify the numbers for qmin/qmax and don't need to use 'reduce_range'.  @jerryzh168

**Comment by jerryzh168** [comment]
yeah, and we should remove reduce_range in our calls as well, when we specify quant_min/quant_max, maybe we can add an assert saying quant_min/quant_max and reduce_range should not be specified at the same time

File: test/quantization/core/test_workflow_module.py [hunk 1]
```diff
@@ -1154,7 +1154,7 @@ def test_fused_mod_per_channel(self):
             torch.testing.assert_allclose(mod.state_dict()['activation_post_process.max_val'], running_max_op)
 
     def test_fused_mod_reduce_range(self):
-        obs = FusedMovingAvgObsFakeQuantize(quant_min=0, quant_max=255, dtype=torch.quint8, reduce_range=True)
+        obs = FusedMovingAvgObsFakeQuantize(quant_min=0, quant_max=127, dtype=torch.quint8, reduce_range=True)
```
Comment (jerryzh168) [inline test/quantization/core/test_workflow_module.py:1157]: we should remove reduce_range here

File: torch/ao/quantization/fake_quantize.py [hunk 2]
```diff
@@ -333,7 +333,7 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
             self.is_symmetric_quant,
         )
 
-default_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=255,
+default_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=127,
```
Comment (jerryzh168) [inline torch/ao/quantization/fake_quantize.py:336]: also here

**Comment by jerryzh168** [comment]
I think we should add an assert to catch everything and make this a bc-breaking change

**Comment by terrychenism** [comment]
> yeah, and we should remove reduce_range in our calls as well, when we specify quant_min/quant_max, maybe we can add an assert saying quant_min/quant_max and reduce_range should not be specified at the same time

Yes makes sense

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35452674).

**Comment by terrychenism** [comment]
> I think a change like this needs a significantly more comprehensive PR summary. It should detail:
> 
> 1. how things work now - including all the edge cases
> 2. what specifically this PR is changing for the user, which users are affected and what is the change for them
> 3. if there are any BC breaking changes, what the users need to change should be laid out

summary updated.

**Comment by vkuzo** [comment]
Hmm, I'm still not sure about the goal.  Before this PR, we had qmin, qmax, reduce_range, which all worked correctly with each other.  It was slightly confusing if reduce_range interacted with qmin and qmax, but it still worked logically.

Now, we still have qmin, qmax and reduce_range, but we've made some of the behaviors which were working correctly before throw an error.  Not sure why?  Could we throw a warning instead of breaking things?
 
I feel like it would make more sense to remove reduce_range from the API, but that's something which will break BC and will need a warning with a specific release version before we do it.

**Comment by jerryzh168** [comment]
> Hmm, I'm still not sure about the goal. Before this PR, we had qmin, qmax, reduce_range, which all worked correctly with each other. It was slightly confusing if reduce_range interacted with qmin and qmax, but it still worked logically.

I think the goal is that we'd like to move away from using both quant_min/quant_max and reduce_range because it does not make a lot of sense, i.e. when user is providing custom quant_min/quant_max, why not just provide the final quant_min/quant_max that is going to be used, but relying on reduce_range to do the divide by 2 calculation?
Yeah and agree that we should just remove reduce_range.

regarding how to get to the final state, if we can't break bc now then I feel it would make sense to put up a warning for now and still support this behavior

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35452674).

**Pull request closed without merging**